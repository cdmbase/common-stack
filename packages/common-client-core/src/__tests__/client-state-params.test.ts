
import gql from 'graphql-tag';
import { TestFeature } from './test-feature';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloClient } from 'apollo-client';
import { Observable, ApolloLink, execute } from 'apollo-link';
import { merge } from 'lodash';

describe('client-state-params', () => {
  const schema = `
  type Query {}
  type Mutation {}
  `;
  const resolvers = { Query: { foo: () => ({ bar: true }) } };
  const defaults = { foo: { bar: true, __typename: 'Bar' } };
  const query = gql`
    {
      foo @client {
        bar
      }
    }
  `;

  const remoteQuery = gql`
    {
      foo {
        bar
      }
    }
  `;

  const typeDefs = `
    type Todo {
      id: String
      message: String!
    }
    type Query {
      todo(id: String!): Todo
    }
  `;


  // feature1
  const resolvers1 = { Query: { foo1: () => ({ bar1: true, __typename: 'Bar1' }) } };
  const defaults1 = { foo1: { bar1: false, __typename: 'Bar1' } };
  const query1 = gql`
      {
        foo1 @client {
          bar1
        }
      }
    `;

  const remoteQuery1 = gql`
      {
        foo1 {
          bar1
        }
      }
    `;

  const typeDefs1 = `
      type Bar1 {
        bar1: Boolean
      }
      extend type Query {
        foo1(): Bar1
      }
    `;

  // feature2
  const resolvers2 = { Query: { foo2: () => ({ bar2: true }) } };
  const defaults2 = { foo2: { bar2: false, __typename: 'Bar2' } };
  const query2 = gql`
      {
        foo2 @client {
          bar2
        }
      }
    `;

  const remoteQuery2 = gql`
      {
        foo2 {
          bar2
        }
      }
    `;

  const typeDefs2 = `
      type Todo2 {
        id: String
        message: String!
      }
      extend type Query {
        todo2(id: String!): Todo2
      }
    `;

  const module1 = new TestFeature({
    clientStateParams: [{ resolvers: resolvers1, defaults: defaults1, typeDefs: typeDefs1 }],
  });

  const module2 = new TestFeature({
    clientStateParams: [{ resolvers: resolvers2, defaults: defaults2, typeDefs: typeDefs2 }],
  });

  const finalModule = new TestFeature(module1, module2, new TestFeature({ clientStateParams: [{ typeDefs: schema }] }));

  it('merge client-state', () => {
    expect({ ...finalModule.getStateParams }).toMatchSnapshot();
  });

  it('writes defaults to the cache upon initialization', () => {
    const cache = new InMemoryCache();

    cache.writeData({
      data: defaults,
    });
    expect(cache.extract()).toMatchSnapshot();
  });

  it('should return empty string when typedefs not defined', () => {
    const moduleNoTypedef = new TestFeature({
      clientStateParams: [{ resolvers: resolvers1, defaults: defaults1 }],
    });

    expect(moduleNoTypedef.getStateParams.typeDefs).toBe('');
  });

  it('concatenates schema strings if typeDefs are passed in as an array', () => {
    const anotherSchema = `
      type Foo {
        foo: String!
        bar: String
      }
    `;

    const nextLink = new ApolloLink(operation => {
      const { schemas } = operation.getContext();
      expect(schemas).toMatchSnapshot();
      return Observable.of({
        data: { foo1: { bar1: true, __typename: 'Bar1' } },
      });
    });

    const cache = new InMemoryCache();
    cache.writeData({
      data: finalModule.getStateParams.defaults,
    });
    const client = new ApolloClient({
      cache,
      resolvers: finalModule.getStateParams.resolvers,
      typeDefs:  schema.concat(finalModule.getStateParams.typeDefs as string),
    });


    client
      .query({ query: query1 })
      .then(({ data }) => {
        expect(data).toMatchObject({ foo1: { bar1: true, __typename: 'Bar1' } });
      });
  });

});
