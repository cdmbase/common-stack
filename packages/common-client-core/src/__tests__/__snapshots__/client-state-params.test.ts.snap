// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client-state-params merge client-state 1`] = `
Object {
  "defaults": Object {
    "foo1": Object {
      "__typename": "Bar1",
      "bar1": false,
    },
    "foo2": Object {
      "__typename": "Bar2",
      "bar2": false,
    },
  },
  "fragmentMatcher": Object {},
  "resolvers": Object {
    "Query": Object {
      "foo1": [Function],
      "foo2": [Function],
    },
  },
  "retryLinkAttemptFuncs": Array [
    [Function],
    [Function],
  ],
  "typeDefs": "type Bar1 {
        bar1: Boolean
      }
      extend type Query {
        foo1(): Bar1
      }
type Todo2 {
        id: String
        message: String!
      }
      extend type Query {
        todo2(id: String!): Todo2
      }
type Query {}
  type Mutation {}",
}
`;

exports[`client-state-params writes defaults to the cache upon initialization 1`] = `
Object {
  "$ROOT_QUERY.foo": Object {
    "__typename": "Bar",
    "bar": true,
  },
  "ROOT_QUERY": Object {
    "foo": Object {
      "generated": true,
      "id": "$ROOT_QUERY.foo",
      "type": "id",
      "typename": "Bar",
    },
  },
}
`;
